File: BraintreeCore/src/main/java/com/braintreepayments/api/ConfigurationCache.java
package com.braintreepayments.api;

import android.content.Context;

import androidx.annotation.VisibleForTesting;

import java.util.concurrent.TimeUnit;

class ConfigurationCache {

    private static final long TIME_TO_LIVE = TimeUnit.MINUTES.toMillis(5);

    private static volatile ConfigurationCache INSTANCE;

    private final BraintreeSharedPreferences sharedPreferences;

    static ConfigurationCache getInstance(Context context) {
        if (INSTANCE == null) {
            synchronized (ConfigurationCache.class) {
                // double check that instance was not created in another thread
                if (INSTANCE == null) {
                    INSTANCE = new ConfigurationCache(BraintreeSharedPreferences.getInstance(context));
                }
            }
        }
        return INSTANCE;
    }

    @VisibleForTesting
    ConfigurationCache(BraintreeSharedPreferences sharedPreferences) {
        this.sharedPreferences = sharedPreferences;
    }

    String getConfiguration(String cacheKey) throws BraintreeSharedPreferencesException {
        return getConfiguration(cacheKey, System.currentTimeMillis());
    }

    @VisibleForTesting
    String getConfiguration(String cacheKey, long currentTimeMillis) throws BraintreeSharedPreferencesException {
        String timestampKey = cacheKey + "_timestamp";
        if (sharedPreferences.containsKey(timestampKey)) {
            long timeInCache = (currentTimeMillis - sharedPreferences.getLong(timestampKey));
            if (timeInCache < TIME_TO_LIVE) {
                return sharedPreferences.getString(cacheKey, "");
            }
        }
        return null;
    }

    void saveConfiguration(Configuration configuration, String cacheKey) throws BraintreeSharedPreferencesException {
        saveConfiguration(configuration, cacheKey, System.currentTimeMillis());
    }

    @VisibleForTesting
    void saveConfiguration(Configuration configuration, String cacheKey, long currentTimeMillis) throws BraintreeSharedPreferencesException {
        String timestampKey = String.format("%s_timestamp", cacheKey);
        sharedPreferences.putStringAndLong(cacheKey, configuration.toJson(), timestampKey, currentTimeMillis);
    }
}

File: BraintreeCore/src/main/java/com/braintreepayments/api/ConfigurationCache.kt
package com.braintreepayments.api

import android.content.Context
import androidx.annotation.VisibleForTesting
import java.util.concurrent.TimeUnit

internal class ConfigurationCache @VisibleForTesting constructor(
        private val sharedPreferences: BraintreeSharedPreferences
    ) {

    fun getConfiguration(cacheKey: String): String? {
        return getConfiguration(cacheKey, System.currentTimeMillis())
    }

    fun getConfiguration(cacheKey: String, currentTimeMillis: Long): String? {
        val timestampKey = "${cacheKey}_timestamp"
        if (sharedPreferences.containsKey(timestampKey)) {
            val timeInCache = currentTimeMillis - sharedPreferences.getLong(timestampKey)
            if (timeInCache < TIME_TO_LIVE) {
                return sharedPreferences.getString(cacheKey, "")
            }
        }
        return null
    }

    fun saveConfiguration(configuration: Configuration, cacheKey: String?) {
        saveConfiguration(configuration, cacheKey, System.currentTimeMillis())
    }

    fun saveConfiguration(
        configuration: Configuration,
        cacheKey: String?,
        currentTimeMillis: Long
    ) {
        val timestampKey = "${cacheKey}_timestamp"
        sharedPreferences.putStringAndLong(
            cacheKey,
            configuration.toJson(),
            timestampKey,
            currentTimeMillis
        )
    }

    companion object {
        private val TIME_TO_LIVE = TimeUnit.MINUTES.toMillis(5)

        @Volatile
        private var INSTANCE: ConfigurationCache? = null
        fun getInstance(context: Context): ConfigurationCache =
            INSTANCE ?: synchronized(this) {
                INSTANCE ?: ConfigurationCache(
                    BraintreeSharedPreferences.getInstance(context)
                ).also { INSTANCE = it }
            }
    }
}
